### PR TITLE
Implementation of Timeline RFC

### PR DESCRIPTION
**Pull request by tsherif**
Updates to timeline base on RFC discussion (#1088)

Updated tests and added an example that demonstrates using multiple coordinated timeline channels.

### CODE REVIEW

File: modules/core/src/lib/timeline.js [hunk 1]
```diff
@@ -101,11 +76,16 @@ export class Timeline {
   }
 
   _setChannelTime(channel, time) {
-    channel.time = time * channel.rate;
-    if (channel.wrapMode === WRAP_LOOP) {
-      channel.time %= channel.duration;
+    const offsetTime = time - channel.delay;
+    const totalDuration = channel.duration * channel.repeat;
+    // Note(Tarek): Don't loop on final repeat.
+    if (offsetTime >= totalDuration) {
```
Comment (Pessimistress) [inline modules/core/src/lib/timeline.js]: Nit: separate out the rate logic for readability

```js
    if (offsetTime < 0) {
      channel.time = 0;
    } else if (offsetTime >= totalDuration) {
      channel.time = channel.duration;
    } else {
      channel.time = offsetTime % channel.duration;
    }
    channel.time *= channel.rate;
```

**Comment by georgios-uber** [comment]
Where is the integration with gltf animator?

**Comment by tsherif** [comment]
@georgios-uber Done.

**Comment by tsherif** [comment]
@Pessimistress docs updated and channel time calculation simplified.

File: modules/addons/src/gltf/gltf-instantiator.js [hunk 2]
```diff
@@ -38,7 +38,7 @@ export default class GLTFInstantiator {
 
   createAnimator() {
     if (Array.isArray(this.gltf.animations)) {
-      return new GLTFAnimator(this.gltf);
+      return new GLTFAnimator(this.gltf, this.options.timeline, this.options.channel);
```
Comment (Pessimistress) [inline modules/addons/src/gltf/gltf-instantiator.js]: - Channels need to be removed with the glTF scene is removed
- Separate out the glTF stuff to a different PR?

Comment (tsherif) [reply modules/addons/src/gltf/gltf-instantiator.js]: The channel isn't added by the animator, so it shouldn't remove it.

File: examples/core/gltf/app.js [hunk 3]
```diff
@@ -262,7 +262,11 @@ export default class AppAnimationLoop extends AnimationLoop {
       blend: false
     });
 
-    this.loadOptions = Object.assign({}, DEFAULT_OPTIONS);
+    const channel = this.timeline.addChannel({
+      rate: 0.5
```
Comment (georgios-uber) [inline examples/core/gltf/app.js]: Why not 1.0

Comment (tsherif) [reply examples/core/gltf/app.js]: Just to have it do something.

File: modules/addons/src/gltf/gltf-instantiator.js (hunk 2 above)
Comment (georgios-uber) [reply modules/addons/src/gltf/gltf-instantiator.js]: Separate out the glTF stuff to a different PR

File: examples/core/gltf/app.js (hunk 3 above)
Comment (georgios-uber) [reply examples/core/gltf/app.js]: ok

**Comment by tsherif** [comment]
@georgios-uber I've already added the gltf integration based on your earlier request, and it's light enough that I don't think it's worth it now to revert and create a separate PR. 

The channel is just a handle that the timeline uses. It's optional and if omitted the base timeline time will be used.

File: modules/core/src/lib/timeline.js [hunk 4]
```diff
@@ -53,30 +44,14 @@ export class Timeline {
   }
 
   setTime(time) {
-    this._setChannelTime(this, time);
+    this.time = Math.max(0, time);
```
Comment (rolyatmax) [inline modules/core/src/lib/timeline.js]: I wonder if this should throw if `time < 0`. If someone passes a negative time they either (1) don't understand `setTime`'s requirements and would benefit from the message, or (2) didn't expect their code to pass a negative number and would benefit from the message. If throwing is too aggressive, then maybe a console warning would do the job?

Comment (tsherif) [reply modules/core/src/lib/timeline.js]: I don't really consider negative input to be an error. It's just that it will be clamped to 0. If negative values are passed in unintentionally, I'm thinking it should be pretty easy to spot, since the time values won't move (e.g. an animation will stop progressing).
